dds/DCPS/Observer.h

```diff
@@ -0,0 +1,122 @@
+/*
+ *
+ *
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#ifndef OPENDDS_DCPS_OBSERVER_H
+#define OPENDDS_DCPS_OBSERVER_H
+
+#include "RcObject.h"
+#include "Definitions.h"
+#include "SequenceNumber.h"
+
+#include <dds/DdsDcpsCoreC.h>
+#include <dds/DdsDcpsPublicationC.h>
+#include <dds/DdsDcpsSubscriptionC.h>
+#include <dds/DdsDcpsInfoUtilsC.h>
+
+#if !defined (ACE_LACKS_PRAGMA_ONCE)
+#pragma once
+#endif /* ACE_LACKS_PRAGMA_ONCE */
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace DCPS {
+
+class DataSampleElement;
+class ReceivedDataSample;
+class ReceivedDataElement;
+
+class OpenDDS_Dcps_Export Observer
+  : public virtual RcObject
+{
+public:
+  typedef Observer* Ptr;
+  typedef TAO_Objref_Var_T<Observer> Var;
+  typedef TAO_Objref_Out_T<Observer> Out;
+  typedef RcHandle<Observer> Rch;
```

Comment:
Is there another example of this style of typedef in the codebase?

Reply:
I did not see in our codebase yet, but the idea is consistently used in the standard library, and I used it for many years for defining custom types, iterators, exceptions, etc.

Reply:
This doesn't seem to be resolved in the latest code.

Reply:
The first three (Ptr, Var, Out) were deleted based on your another comment above. Rch is used. As for this style of typedef in the codebase, it's true that there are so many typedef's in the style of ClassName_var, but there are actually this style of typedef's in class declaration, too. For example, the TypeSupportVar in the original tests/DCPS/DataRepresentation/DataRepresentation.cpp (and the _var_type in TypeSupportType):

template<typename Type>
class RegisteredType {
public:
  typedef typename OpenDDS::DCPS::DDSTraits<Type> TypeTraits;
  typedef typename TypeTraits::TypeSupportType TypeSupport;
  typedef typename TypeSupport::_var_type TypeSupportVar;
  typedef typename TypeTraits::TypeSupportTypeImpl TypeSupportImpl;

  TypeSupportVar type_support_var_;

  RegisteredType()
  : type_support_var_(0)
  {
  }
//...

As long as no spec requires otherwise, I would like to suggest using this in-class-declaration style, since it's consistent with the standard library and with the more recent C++ practice. (e.g. before STL, an iterator of ClassX was most likely to be defined as class ClassX_Iterator; now it's simple ClassX::iterator.)

I hope the in-class typedef RcHandle<Observer> Rch; is acceptable here. 
If not, I can replace it with outside-class typedef RcHandle<Observer> Observer_rch; and replace all the use of Observer::Rch with Observer_rch.

Reply:
We should follow the established patterns for how RcObject/RcHandle are used.